### ISSUE TITLE
return inside a for loop causes stack misalignment

### ISSUE THREAD

**Issue by abenkovskii** [KSP-KOS/KOS#756]
![screenshot4](https://cloud.githubusercontent.com/assets/11031560/7048626/dc9ed9ba-de1c-11e4-8a02-b440602cbd3b.png)

**Comment by abenkovskii** [comment]
log: https://www.dropbox.com/s/knv9rmch7bs9sjn/output_log-return_bug.txt?dl=0

**Comment by Dunbaratu** [comment]
Ah crud.  I have a list of these to go over after the closure and local lock, renames are done.  You've done a good job finding the edge cases.

**Comment by abenkovskii** [comment]
thanks!

**Comment by Dunbaratu** [comment]
Well I can see what is wrong from the log file (thank you for that).  This line here:

1/bug                  12:14  0054 popscope 1 

Should have been a popscope 3.  (pop 3 scopes from the stack).  The compiler is supposed to put the correct number in there "automatically" based on counting the diff between the current nest level and what the nest level was when the function started.  It worked when I tested it before.  I wonder if I broke something recently.

Anyway thank you for the detailed report.  I won't be able to look at the C# code until late tonight.

**Comment by hvacengi** [comment]
If it helps with debugging, the error also seems to occur within `if { ... } else { ... }` blocks.  I get the same error (and a `popscope  1` before the return) by running the following code:
```
declare function ag1stat {
	if ag1 {
		return "true".
	}
	else {
		return "false".
	}
}
print ag1stat().
```
I tested to see if there was an issue using multiple returns in nested functions (ie func1 returns a value based on func2) and so no problems.  I took a quick look at what I think is the applicable C#, and it appears to be trying to adjust the popscope number at the very least, but it must be missing something.

**Comment by hvacengi** [comment]
More importantly... I just read his existing code and saw I'm a duplicate.  Disregard my post.  I don't know what I though the original code was doing...

**Comment by Dunbaratu** [comment]
Should get resolved by what was in #741 as it had to be merged with a few other things because I screwed up git a bit.

**Comment by Dunbaratu** [comment]
I tested a large use case like this with plenty of nesting and after the 741 fixes it was working.
